dialogue-core/src/main/java/com/palantir/dialogue/core/StickyEndpointChannels.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.core;
+
+import com.github.benmanes.caffeine.cache.Ticker;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Suppliers;
+import com.google.common.collect.ImmutableList;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.palantir.dialogue.Channel;
+import com.palantir.dialogue.Endpoint;
+import com.palantir.dialogue.EndpointChannel;
+import com.palantir.dialogue.EndpointChannelFactory;
+import com.palantir.dialogue.Request;
+import com.palantir.dialogue.Response;
+import com.palantir.dialogue.core.BalancedScoreTracker.ChannelScoreInfo;
+import com.palantir.logsafe.Preconditions;
+import com.palantir.random.SafeThreadLocalRandom;
+import com.palantir.tritium.metrics.registry.TaggedMetricRegistry;
+import java.util.List;
+import java.util.Random;
+import java.util.function.Supplier;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.ThreadSafe;
+
+/** Allows requests for a 'transaction' should land on one host. */
+public final class StickyEndpointChannels {
+
+    private final ImmutableList<? extends EndpointChannelFactory> channels;
+    private final BalancedScoreTracker tracker;
+
+    private StickyEndpointChannels(Builder params) {
+        this.channels = params.channels;
+        // TODO(dfox): handle 'channels' of size 0/1 (balanced tracker will barf)
```

Comment:
Can we replace this todo with a preconditions check, or potentially move handling for the zero/one channel cases into Builder.build?

Reply:
Ah I should delete this. BalancedScoreTracker already has a `Preconditions.checkState(channelCount >= 2, "At least two channels required");`, and in ReloadingClientFactory we avoid even constructing one of these if we detect 0 or 1 uris.